pkg/koordlet/runtimehooks/hooks/cpunormalization/cpu_normalization.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (p *Plugin) adjustPodCPUShares(podCtx *protocol.PodContext) error {
+	if !p.rule.IsEnabled() {
+		return nil
+	}
+
+	originalCPUShares := podCtx.Request.Resources.CPUShares
+	if originalCPUShares == nil || *originalCPUShares <= 0 { // no need to scale when cgroup is unset
+		return nil
+	}
+
+	ratio := p.rule.GetCPUNormalizationRatio()
+	if ratio == 1.0 { // no need to change
+		podCtx.Response.Resources.CPUShares = originalCPUShares
+		return nil
+	}
+
+	cpuShares := int64(math.Ceil(float64(*originalCPUShares) / ratio))
+	podCtx.Response.Resources.CPUShares = &cpuShares
+	klog.V(6).Infof("plugin %s adjusts pod %s/%s cpu shares from %d to %d",
+		name, podCtx.Request.PodMeta.Namespace, podCtx.Request.PodMeta.Name, *originalCPUShares, cpuShares)
+
+	return nil
+}
+
+func (p *Plugin) adjustPodCFSQuota(podCtx *protocol.PodContext) error {
+	if !p.rule.IsEnabled() {
+		return nil
+	}
+
+	originalCFSQuota := podCtx.Request.Resources.CFSQuota
+	if originalCFSQuota == nil || *originalCFSQuota <= 0 { // no need to scale when cgroup is unset
+		return nil
+	}
+
+	ratio := p.rule.GetCPUNormalizationRatio()
+	if ratio == 1.0 { // no need to change
+		podCtx.Response.Resources.CFSQuota = originalCFSQuota
+		return nil
+	}
+
+	cfsQuota := int64(math.Ceil(float64(*originalCFSQuota) / ratio))
+	podCtx.Response.Resources.CFSQuota = &cfsQuota
+	klog.V(6).Infof("plugin %s adjusts pod %s/%s cfs quota from %d to %d",
+		name, podCtx.Request.PodMeta.Namespace, podCtx.Request.PodMeta.Name, *originalCFSQuota, cfsQuota)
+
+	return nil
+}
+
+func (p *Plugin) adjustContainerCPUShares(containerCtx *protocol.ContainerContext) error {
+	if !p.rule.IsEnabled() {
+		return nil
+	}
+
+	originalCPUShares := containerCtx.Request.Resources.CPUShares
+	if originalCPUShares == nil || *originalCPUShares <= 0 { // no need to scale when cgroup is unset
+		return nil
+	}
+
+	ratio := p.rule.GetCPUNormalizationRatio()
+	if ratio == 1.0 { // no need to change
+		containerCtx.Response.Resources.CPUShares = originalCPUShares
+		return nil
+	}
+
+	cpuShares := int64(math.Ceil(float64(*originalCPUShares) / ratio))
+	containerCtx.Response.Resources.CPUShares = &cpuShares
+	klog.V(6).Infof("plugin %s adjusts container %s/%s/%s cpu shares from %d to %d", name,
+		containerCtx.Request.PodMeta.Namespace, containerCtx.Request.PodMeta.Name,
+		containerCtx.Request.ContainerMeta.Name, *originalCPUShares, cpuShares)
+
+	return nil
+}
+
+func (p *Plugin) adjustContainerCFSQuota(containerCtx *protocol.ContainerContext) error {
+	if !p.rule.IsEnabled() {
+		return nil
+	}
+
+	originalCFSQuota := containerCtx.Request.Resources.CFSQuota
+	if originalCFSQuota == nil || *originalCFSQuota <= 0 { // no need to scale when cgroup is unset
+		return nil
+	}
+
+	ratio := p.rule.GetCPUNormalizationRatio()
+	if ratio == 1.0 { // no need to change
+		containerCtx.Response.Resources.CFSQuota = originalCFSQuota
+		return nil
+	}
+
+	cfsQuota := int64(math.Ceil(float64(*originalCFSQuota) / ratio))
```

Comment:
Is it possible for ratio to be 0？

Reply:
possible when the user manually updates the ratio. fixed by adjust only if the ratio is lager than 1